File: Daily_Log/Singletons/CBDatabase.m
//
//  CBUtils.m
//  Daily Log
//
//  Created by Ayush Sood on 10/15/14.
//  Copyright (c) 2014 Kanler Inc. All rights reserved.
//

#import "CBDatabase.h"
#import "AFHTTPRequestOperationManager.h"
#import "UICKeyChainStore.h"

NSString *const API_URL = @"http://kanler.com/api/v1/";

@implementation CBDatabase

+ (CBDatabase *)sharedInstance {
    static CBDatabase *sharedDB = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedDB = [[self alloc] init];
    });
    return sharedDB;
}

- (void)authUserWithEmail:(NSString *)email password:(NSString *)password callback:(SEL)callback delegate:(id)delegate
{
    NSString *url = @"auth/";
    NSDictionary *data = @{
                           @"user_email": email,
                           @"user_password": password
                           };
    
    [self makePOSTRequestToEndpoint:url withData:data withCallback:callback withDelegate:delegate];
}

- (void)fetchProjectsWithCallback:(SEL)callback delegate:(id)delegate {
    NSString *url = @"logging/projects/";
    [self makeGETRequestToEndpoint:url withCallback:callback withDelegate:delegate];
}

- (void)updateProject:(NSNumber *)projectId data:(NSDictionary *)data callback:(SEL)callback delegate:(id)delegate {
    NSString *url = [@"logging/projects/" stringByAppendingString:[projectId stringValue]];
    [self makePUTRequestToEndpoint:url withData:data withCallback:callback withDelegate:delegate];
}

- (void)addLogToProject:(NSNumber *)projectId data:(NSDictionary *)data callback:(SEL)callback delegate:(id)delegate
{
    NSString *url = [@"logging/projects/" stringByAppendingString:@""];
    [self makePOSTRequestToEndpoint:url withData:data withCallback:callback withDelegate:delegate];
}

- (void)makeGETRequestToEndpoint:(NSString *)url withCallback:(SEL)callback withDelegate:(id)delegate {
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    UICKeyChainStore *store = [UICKeyChainStore keyChainStore];
    if (store[@"token"] != nil) {
        [manager.requestSerializer setValue:store[@"token"] forHTTPHeaderField:@"X-USER-TOKEN"];
        [manager.requestSerializer setValue:store[@"email"] forHTTPHeaderField:@"X-USER-EMAIL"];
    }
    
    [manager GET:[API_URL stringByAppendingString:url] parameters:nil success:^(AFHTTPRequestOperation *operation, id responseObject) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        NSLog(@"response = %@", responseObject);
        [delegate performSelector:callback withObject:responseObject];
#pragma clang diagnostic pop
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        //
        NSLog(@"%@", error);
    }];
}

- (void)makePOSTRequestToEndpoint:(NSString *)url withData:(NSDictionary *)data withCallback:(SEL)callback withDelegate:(id)delegate
{
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    UICKeyChainStore *store = [UICKeyChainStore keyChainStore];
    if (store[@"token"] != nil) {
        [manager.requestSerializer setValue:store[@"token"] forHTTPHeaderField:@"X-USER-TOKEN"];
        [manager.requestSerializer setValue:store[@"email"] forHTTPHeaderField:@"X-USER-EMAIL"];
    }
    
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:data
                                                       options:NSJSONWritingPrettyPrinted // Pass 0 if you don't care about the readability of the generated string
                                                         error:&error];
    NSString *jsonString = @"";
    if (! jsonData) {
        NSLog(@"Got an error: %@", error);
    } else {
        jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        NSLog(@"json string :%@", jsonString);
    }
    
    NSData *objectData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:objectData
                                                         options:NSJSONReadingMutableContainers
                                                           error:&error];
    
    [manager POST:[API_URL stringByAppendingString:url] parameters:json success:^(AFHTTPRequestOperation *operation, id responseObject) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [delegate performSelector:callback withObject:@{@"status": [NSNumber numberWithInteger:[operation.response statusCode]], @"response": responseObject}];
#pragma clang diagnostic pop
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        NSLog(@"error: %@", error);
        [delegate performSelector:callback withObject:@{@"status":[NSNumber numberWithInteger:[operation.response statusCode]], @"response": error}];
#pragma clang diagnostic pop
    }];
}

- (void)makePUTRequestToEndpoint:(NSString *)url withData:(NSDictionary *)data withCallback:(SEL)callback withDelegate:(id)delegate
{
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    UICKeyChainStore *store = [UICKeyChainStore keyChainStore];
    if (store[@"token"] != nil) {
        [manager.requestSerializer setValue:store[@"token"] forHTTPHeaderField:@"X-USER-TOKEN"];
        [manager.requestSerializer setValue:store[@"email"] forHTTPHeaderField:@"X-USER-EMAIL"];
    }
    
    [manager PUT:[API_URL stringByAppendingString:url] parameters:data success:^(AFHTTPRequestOperation *operation, id responseObject) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [delegate performSelector:callback withObject:@{@"status": [NSNumber numberWithInteger:[operation.response statusCode]], @"response": responseObject}];
#pragma clang diagnostic pop
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        NSLog(@"error: %@", error);
        [delegate performSelector:callback withObject:@{@"status":[NSNumber numberWithInteger:[operation.response statusCode]], @"response": error}];
#pragma clang diagnostic pop
    }];
}

@end

File: Daily_Log/Singletons/CBDatabase.swift
//
//  CBDatabase.swift
//  Daily_Log
//
//  Created by Tommi on 3/13/15.
//  Copyright (c) 2015 Tommi. All rights reserved.
//

import UIKit

class CBDatabase: NSObject {

    let API_URL = "http://kanler.com/api/v1/"
    
    struct STATICVALUES {
        static var onceToken :  dispatch_once_t = 0
        static var sharedInstance : CBDatabase!
    }
    
    class func sharedInstance() -> CBDatabase!
    {
        dispatch_once(&STATICVALUES.onceToken, { () -> Void in
          STATICVALUES.sharedInstance = CBDatabase()
        })
        
        return STATICVALUES.sharedInstance
    }
    
    func authUserWithEmail(email : String, password : String, callback : Selector, delegate : AnyObject )
    {
        var url = "auth/" as String
        var data = [ "user_email" : email,  "user_password" : password ]  as NSDictionary
        self.makePOSTRequestToEndpoint(url, withData: data, withCallback: callback, withDelegate: delegate)
    }

    func fetchProjectsWithCallback (callback : Selector, delegate : AnyObject)
    {
        var url = "logging/projects/" as String
        self.makeGETRequestToEndpoint(url, withCallback: callback, withDelegate: delegate)
    }

    func updateProject(projectId : NSNumber, data : NSDictionary, callback : Selector, delegate : AnyObject)
    {
        var url = "logging/projects/\(projectId.stringValue)" as String
        self.makePUTRequestToEndpoint(url, withData: data, withCallback:callback, withDelegate: delegate)
    }
    
    func addLogToProject(projectId : NSNumber, data : NSDictionary, callback : Selector, delegate : AnyObject)
    {
        var url = "logging/projects/" as String
        self.makePOSTRequestToEndpoint(url, withData: data, withCallback: callback, withDelegate: delegate)
    }

    func makeGETRequestToEndpoint(url : String, withCallback callback : Selector, withDelegate delegate : AnyObject)
    {
        var manager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFJSONResponseSerializer(readingOptions: NSJSONReadingOptions.allZeros)
        
        var store = UICKeyChainStore()
        if(store["token"] != nil)
        {
            manager.requestSerializer.setValue(store["token"], forHTTPHeaderField: "X-USER-TOKEN")
            manager.requestSerializer.setValue(store["email"], forHTTPHeaderField: "X-USER-EMAIL")
        }
        
        manager.GET(API_URL + url, parameters: nil, success: { (operation : AFHTTPRequestOperation!, responseObject : AnyObject!) -> Void in
                let delay = 0.0
                println("response = \(responseObject)")
                NSTimer.scheduledTimerWithTimeInterval(delay, target: delegate, selector: callback, userInfo: responseObject , repeats: false)
            }) { (operation : AFHTTPRequestOperation!, error : NSError!) -> Void in
                println(error)
        }
    }

    func makePOSTRequestToEndpoint(url : String, withData data : NSDictionary , withCallback callback : Selector, withDelegate delegate : AnyObject)
    {
        var manager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFJSONResponseSerializer(readingOptions: NSJSONReadingOptions.allZeros)
        manager.requestSerializer = AFJSONRequestSerializer(writingOptions: NSJSONWritingOptions.allZeros)
        
        var store = UICKeyChainStore()
        if(store["token"] != nil)
        {
            manager.requestSerializer.setValue(store["token"], forHTTPHeaderField: "X-USER-TOKEN")
            manager.requestSerializer.setValue(store["email"], forHTTPHeaderField: "X-USER-EMAIL")
        }
        
        var error : NSError?
        var jsonData = NSJSONSerialization.dataWithJSONObject(data, options: NSJSONWritingOptions.PrettyPrinted, error: &error)
        var jsonString : NSString! = ""
        if( jsonData == nil )
        {
            println("Got an error: \(error)")
        }
        else
        {
            jsonString = NSString(data: jsonData!, encoding: NSUTF8StringEncoding)
            println("json string : \(jsonString)")
        }
        
        var objectData = jsonString.dataUsingEncoding(NSUTF8StringEncoding) as NSData!
        var json = NSJSONSerialization.JSONObjectWithData(objectData, options: NSJSONReadingOptions.MutableContainers, error: &error) as NSDictionary?
        
        manager.POST(API_URL + url, parameters: json, success: { (operation : AFHTTPRequestOperation!, responseObject : AnyObject!) -> Void in
            let delay = 0.0
            NSTimer.scheduledTimerWithTimeInterval(delay, target: delegate, selector: callback, userInfo: ["status" : NSNumber(integer:(operation.response.statusCode)), "response" : responseObject ], repeats: false)

            }) { (operation : AFHTTPRequestOperation!, error :NSError!) -> Void in
                let delay = 0.0
                println("error = \(error)")
                NSTimer.scheduledTimerWithTimeInterval(delay, target: delegate, selector: callback, userInfo: ["status" : NSNumber(integer:(operation.response.statusCode)), "response" : error ], repeats: false)
        }
        
    }
    
    func makePUTRequestToEndpoint(url : String, withData data : NSDictionary , withCallback callback : Selector, withDelegate delegate : AnyObject)
    {
        var manager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFJSONResponseSerializer(readingOptions: NSJSONReadingOptions.allZeros)
        manager.requestSerializer = AFJSONRequestSerializer(writingOptions: NSJSONWritingOptions.allZeros)
        
        var store = UICKeyChainStore()
        if(store["token"] != nil)
        {
            manager.requestSerializer.setValue(store["token"], forHTTPHeaderField: "X-USER-TOKEN")
            manager.requestSerializer.setValue(store["email"], forHTTPHeaderField: "X-USER-EMAIL")
        }
        
        manager.PUT(API_URL + url, parameters: data, success: { (operation : AFHTTPRequestOperation!, responseObject : AnyObject!) -> Void in
                let delay = 0.0
                NSTimer.scheduledTimerWithTimeInterval(delay, target: delegate, selector: callback, userInfo: ["status" : NSNumber(integer:(operation.response.statusCode)), "response" : responseObject ], repeats: false)
            
            }) { (operation : AFHTTPRequestOperation!, error : NSError!) -> Void in
                let delay = 0.0
                NSTimer.scheduledTimerWithTimeInterval(delay, target: delegate, selector: callback, userInfo: ["status" : NSNumber(integer:(operation.response.statusCode)), "response" : error ], repeats: false)
        }
    }
}


